replace with new tn-models-fp

paribaker opened the pull request:
## What this does

Moves vue js to the new tn-models-fp library



## How to test
Sign up to app
login to app
Start forgot password flow (note emails are not set up)

paribaker commented on {{cookiecutter.project_slug}}/clients/web/vue3/src/services/AxiosClient.js:
```diff
@@ -18,7 +18,7 @@ class ApiService {
   constructor() {
-    let baseURL = `${window.location.protocol}//${window.location.host}/api`
+    let baseURL = `${window.location.protocol}//${window.location.host}/`
 
```
Thoughts on this? We do not append the `api` to the domain in the react side, I think that is a better approach since it is more explicit and controlled by the dev

oudeismetis replied:
My thought would be the opposite. The AxiosClient is making all of it's calls to the backend API.
The baseURL for the backend API starts at `/api`.

mayakeeley replied:
I'm with Ed on this one, I'd prefer /api stayed there

paribaker replied:
sounds good will update appropriately

paribaker commented on {{cookiecutter.project_slug}}/clients/web/vue3/src/services/base-model.js:
```diff
@@ -0,0 +1,8 @@
+import { z } from 'zod'
+import {readonly} from '@thinknimble/tn-models-fp'
+
```
@lakardion getting a zod error here that id is required when using userApi.create(), I temporarily set it to `.optional` to get it working, but I thought that by making it readonly it would not be an issue?

lakardion replied:
sounds like a bug. but at the same time you need not to make id readonly. it is readonly by default

lakardion replied:
Tracking here. As a workaround, not marking `id` as readonly it should work!
https://github.com/thinknimble/tn-models-fp/issues/132

paribaker replied:
okay will leave this as is for now

lakardion replied:
I think there's a bug with the readonly implementation when used on `id`, due to its duality with being readonly regardless being marked as readonly or not. 
readonly should be fine for other fields, until I get to the aforementioned issue we should probably leave `id` without the `readonly` wrap

paribaker replied:
Okay will do, regardless the issue was occurring before I added the readonly, I was just trying to fix it by using the readonly method

lakardion replied:
Okay I didn't notice the create shape override. Yeah readonly is not supported in the override models. Will fix that so that they're completely ignored.

lakardion replied:
Upgrade to `2.12.3` and lmk how it goes

https://github.com/thinknimble/tn-models-fp/pull/137

oudeismetis commented on {{cookiecutter.project_slug}}/clients/mobile/react-native/src/services/user/api.ts:
```diff
@@ -37,7 +38,7 @@ const resetPassword = createCustomServiceCall(
   {
     client: axiosInstance,
-    baseUri: '/api/users/',
+    baseUri: 'api/users/',
```
fun

lakardion replied:
Hmm I don't think that slash matters given that the axios instance should already have the baseUrl defined?, so whether or not you add the slash, axios will know whether it should add it or not

oudeismetis replied:
I was just observing that when I've seen changes like this in the past, it's because things are fragile with slashes missing or extra ones sneaking in and it being easy for someone to break.
Ideally we have a URL builder or a utility that sanitizes URLs before we use them. This exists on the BE in Python as part of the standard library. Not sure if we have it on the frontend.

Not relevant for this PR. Just a code smell.

paribaker replied:
Well actually @lakardion built tn-models to handle that (at least the  trailing slash) right?

lakardion replied:
trailing is handled.
leading, imo doesn't matter. ran a quick test and the request is the same whether you do `api` or `/api` with axios

oudeismetis replied:
nice!

oudeismetis commented on {{cookiecutter.project_slug}}/clients/web/vue3/src/services/users/models.js:
```diff
@@ -1,14 +1,29 @@
 
-import UserAPI from './api'
+export const userShape = {
+  ...baseModelShape,
```
Why are these all "shapes"?

lakardion replied:
https://github.com/thinknimble/tn-models-fp/issues/145

Basically they comply with type from `zod` library. 

We could call them `models` if we wanted to, but they still would need to match the `ZodRawShape` structure.

oudeismetis replied:
gotcha. So we call them shapes because zod calls them shapes. Fair enough.
If it's an internal only zod thing that most consumers of zod wouldn't know about I would consider calling it something else.
Models may or may not be the answer there.

paribaker replied:
Yea I like the idea of shapes over models, but I could also be content with `structure` `entity` `interface` <--- maybe less this one. I think models is a bit incorrect since these aren't true db backed 'models'

oudeismetis replied:
shapes seems fine then

oudeismetis commented on {{cookiecutter.project_slug}}/clients/web/vue3/src/views/ResetPassword.vue:
```diff
@@ -58,7 +58,7 @@ export default {
       const { uid, token } = route.params
 
-      User.api
+      userApi.csc
```
what's `csc`?

lakardion replied:
It's just an alias for `customServiceCall`

https://github.com/thinknimble/tn-models-fp#createcustomservicecall

`.customServiceCall` can still be used if you preferred the longer name.

Ideally all method calls should probably be available at the root of the api object, but the typing for that was quite a pain so any custom call live under `customServiceCall`.

oudeismetis commented on {{cookiecutter.project_slug}}/clients/web/vue3/src/views/Signup.vue:
```diff
@@ -92,8 +93,7 @@ export default {
 
-      User.api
-        .registerUser({
+      userApi.create({
```
`csc` missing here?

lakardion replied:
These are just built-ins so they live at the root of the api object

https://github.com/thinknimble/tn-models-fp#built-in-methods

mayakeeley commented on {{cookiecutter.project_slug}}/clients/web/vue3/src/services/users/api.js:
```diff
@@ -1,81 +1,48 @@
+    inputShape: forgotPasswordShape,
+  },
+  async ({ client, input }) => {
+    console.log(input)
```
Are we deliberately keeping a console.log here? Is there a reason for that?

mayakeeley commented on {{cookiecutter.project_slug}}/clients/mobile/react-native/src/services/user/api.ts:
```diff
@@ -22,6 +22,7 @@ const requestPasswordResetCode = createCustomServiceCall(
     await client.get(`api/password/reset/code/${input.email}/`)
```
Not sure if I've misread this but it looks like all of the other api calls have been changed to remove code. Should this not match that pattern too?

paribaker replied:
ah this is a different endpoint to the one for the web views. In web we use the reset link, but in mobile we use a reset code. We found this approach to be a less cumbersome approach than trying to figure out and get `in-app` links working.

oudeismetis commented on {{cookiecutter.project_slug}}/clients/web/react/src/services/user/api.ts:
```diff
@@ -14,12 +14,12 @@ const login = createCustomServiceCall(
   },
   async ({ client, input }) => {
-    await client.get(`api/password/reset/code/${input.email}/`)
+    await client.get(`api/password/reset/${input.email}/`)
```
a change here but none to the backend?

paribaker replied:
The backend is all set, we probably copied some of this code from an app that was using code only the mobile side needs this code!

oudeismetis commented on {{cookiecutter.project_slug}}/clients/web/vue3/src/services/users/api.js:
```diff
@@ -1,81 +1,47 @@
+    outputShape: userShape,
+  },
+  async ({ client, input, utils }) => {
+    const res = await client.post('api/login/', utils.toApi(input))
```
I think we undid that one change, but all these URLs still start with `/api`. That might be why Cypress is failing

paribaker replied:
yep we made the change but I  thought we said it doesn't matter in the end!

paribaker commented:
@oudeismetis @lakardion okay this is giving me a headache. To resolve once and for all I've added the `/api` to the axios instance on all 3 client/mobile side apps and removed `api` or `/api` from all services folders!

Will test this out on mobile and then check the react side!

oudeismetis commented on .github/workflows/pytest.yml:
```diff
@@ -45,7 +45,7 @@ jobs:
         run: |
           cd my_project
-          pipenv install --dev --deploy --skip-lock
+          pipenv install --dev --deploy
```
why did we make this change?

oudeismetis replied:
yeaaa....pip install couldn't possibly have the --dev flag. Lets merge this as-is and open a github issue to figure out a way to bring this back using pipenv

oudeismetis commented:
> @oudeismetis @lakardion okay this is giving me a headache. To resolve once and for all I've added the `/api` to the axios instance on all 3 client/mobile side apps and removed `api` or `/api` from all services folders!
> 
> Will test this out on mobile and then check the react side!

I appreciate the efforts to keep things standard, though technically it doesn't matter if Vue and React differ a bit in their approach.
For our teams sake though, I'm sure it would be confusing to bounce between 2 projects doing things slightly differently like this.
